api/v1alpha2/machine_types.go

```diff
@@ -52,6 +52,11 @@ type MachineSpec struct {
 	// offered by an infrastructure provider.
 	InfrastructureRef corev1.ObjectReference `json:"infrastructureRef"`
 
+	// InfrastructureTemplateRef is an optional reference to the infrastructure Template resource
+	// that was copied for this machine, if any.
+	// +optional
+	InfrastructureTemplateRef *corev1.ObjectReference `json:"infrastructureTemplateRef,omitempty"`
```

Comment:
```suggestion
	// InfrastructureClonedFrom is an optional reference to the infrastructure template resource
	// that was copied for this machine, if any. 
	// 
	// This value is not meant to be set by users, controllers creating and managing Machine objects
	// may set this value to inform users that the object has been created from a cloned template.
	// +optional
	InfrastructureClonedFrom *corev1.ObjectReference `json:"infrastructureTemplateRef,omitempty"`
```

Clarified a little the wording and text, also renamed it to make sure it's not used for user input.

We might also want to make sure it's immutable by rejecting any changes in validation webhooks.

Reply:
Thinking about this a little more, I would like to get more 👀 and make sure we want to keep this here vs having it on the InfrastructureMachine object itself, although it could get a little tricky to know if the template has this field or not.

If we have it on the InfrastructureMachine, it will also appear on the InfrastructureMachineTemplate, and we could potentially set it in a generic way in the `external.CloneTemplate` function.

Reply:
I agree with @vincepri that this feels a bit off having it here rather than in the provider-defined Template types. I'm not quite sure how we'd validate that an provider-defined type supports a given field to know how to set it (as opposed to the conditional checks we leverage today, which are based on the presence of a field in an unstructured resource).

Reply:
What if we set a simple annotation on the InfrastructureMachine with the ObjectReference (Kind/Version/Name)?

Reply:
I think the annotation approach is fine for the short term, but we should probably at least track trying to find a better way to do it longer term.

To simplify things (especially if we are planning on using an annotation), we can likely omit Kind/Version, since those can be inferred from the source type (the same way that we do during cloning).

Part of me really would like to be able to lean on Conditions here, but existing in Status means we can't rely on it, though.

Reply:
That makes sense, I was thinking in the future when we'll have a "provider registration" or something similar to overcame the need for labels on CRD as well, we can probably make provider expose a list of features/contracts they fulfill and implement.

That would probably also require conformance 🤔.

For now, I'm +1 using json-marshalling from the template ObjectReference to an annotation

Reply:
I lean slightly toward the Machine because and 1) CAPI is the "owner" of the field, in the sense that it's where the clone / template behavior is happening, and 2) the AWSMachine reconciler already looks at the Machine its associated with for the kubernetes version, so it's already got it handy.

I don't imagine it'd be too much harder to have `external.CloneTemplate` just set the field & let the API server drop it if it's unknown. I feel that'd be slightly harder to debug for an infrastructure provider if it wasn't working, but maybe there's something we could do around optional contract annotations & events or so.

Reply:
Oh, my – github didn't update all the comments before I replied. Reading back now.

Reply:
Having on the cloned object would be great to:
- Have a single annotation like `cluster.x-k8s.io/cloned-from` that can be applied on any cloned object.
- Have the annotation to be automatically applied in a generic way in the `external.CloneTemplate` utilities that multiple folks use, they'd get the annotation for free.

Another thing to point out is that the Machine itself isn't templated, or cloned from anything, but only the bootstrap/infrastructure objects are.